pkg/engine/lifecycle_test.go

```diff
@@ -1276,3 +1276,47 @@ func TestCheckFailureInvalidPropertyRecord(t *testing.T) {
 	p.Run(t, nil)
 
 }
+
+// Test that tests that Refresh can detect that resources have been deleted and removes them
+// from the snapshot.
+func TestRefreshWithDelete(t *testing.T) {
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				ReadF: func(
+					urn resource.URN, id resource.ID, props resource.PropertyMap,
+				) (resource.PropertyMap, resource.Status, error) {
+					// This thing doesn't exist.
+					return nil, resource.StatusOK, nil
+				},
+			}, nil
+		}),
+	}
+
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		return nil
+	})
+
+	host := deploytest.NewPluginHost(nil, program, loaders...)
+	p := &TestPlan{Options: UpdateOptions{host: host}}
+
+	provURN := p.NewProviderURN("pkgA", "default", "")
+	urn := p.NewURN("pkgA:m:typA", "resA", "")
+	old := &deploy.Snapshot{
```

Comment:
Nit: rather than explicitly creating an old snapshot, you could run a program that creates resource prior to the refresh. Up to you.

Reply:
Hmm. Doing that is certainly more honest about how people are actually going to run into this situation. I'll think I'll go with that.